torch_geometric/nn/models/graphmask_explainer.py

```diff
@@ -0,0 +1,838 @@
+                        latest_target_embeddings.append(
+                            module.latest_target_embeddings)
+                gate_input = [
+                    latest_source_embeddings, latest_messages,
+                    latest_target_embeddings
+                ]
+                for i in range(self.num_layers):
+                    output = self.full_biases[i]
+                    for j in range(len(gate_input)):
+                        partial = self.gates[i * 4][j](gate_input[j][i])
+                        result = self.gates[(i * 4) + 1][j](partial)
+                        output = output + result
+                    relu_output = self.gates[(i * 4) + 2](output /
+                                                          len(gate_input))
+                    sampling_weights = self.gates[(i * 4) +
+                                                  3](relu_output).squeeze(
+                                                      dim=-1)
+                    sampling_weights, penalty = self.hard_concrete(
+                        sampling_weights)
+                    gates.append(sampling_weights)
+                    total_penalty += penalty
+
+                self.__inject_messages__(gates, self.baselines)
+
+                self.lambda_op = torch.tensor(self.init_lambda,
+                                              requires_grad=True)
+                optimizer_lambda = torch.optim.RMSprop(
+                    [self.lambda_op], lr=self.lambda_optimizer_lr,
+                    centered=True)
+
+                optimizer.zero_grad()
+                optimizer_lambda.zero_grad()
+
+                h = x * self.node_feat_mask.sigmoid()
+                out = self.model(x=h, edge_index=edge_index, batch=batch,
+                                 **kwargs)
+
+                self.__inject_messages__(gates, self.baselines, True)
+
+                if self.return_type == 'regression':
+                    loss = self._loss([-1], out, prediction, total_penalty)
+                else:
+                    log_logits = self._to_log_prob(out)
+                    loss = self._loss([-1], log_logits, prediction,
+                                      total_penalty)
+                loss.backward()
+                optimizer.step()
+                self.lambda_op.grad *= -1
+                optimizer_lambda.step()
+
+                if self.lambda_op.item() < -2:
+                    self.lambda_op.data = torch.full_like(
+                        self.lambda_op.data, -2)
+                elif self.lambda_op.item() > 30:
+                    self.lambda_op.data = torch.full_like(
+                        self.lambda_op.data, 30)
+
+                if self.log:  # pragma: no cover
+                    pbar.update(1)
+
+            if self.log:  # pragma: no cover
+                pbar.close()
+
+        node_feat_mask = self.node_feat_mask.detach().sigmoid().squeeze()
```

Comment:
Reset the explainer stuff inside the original model?

Reply:
I think `GNNExplainer` also implements this same statement (Line 643 above) in a similar way. Please check [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/gnn_explainer.py) on Line 177.